libs/ui/src/virtual_keyboard/virtual_keyboard.tsx

```diff
@@ -206,29 +227,266 @@ export const SPACE_BAR_KEY: Key = {
+// 1. Find the currently focused keyboard button and compute its midpoint
+// 2. In the adjacent row, calculate the closest edge (left or right) for each keyboard button
+// 3. Choose the keyboard button with the closest edge
+function getAdjacentRowTargetButtonIndex(
+  keyMap: KeyMap,
+  rowRefs: React.MutableRefObject<Array<HTMLDivElement | null>>,
+  focusedRowIndex: number,
+  direction: -1 | 1
+) {
+  // The first time the user interacts with the keyboard there is no focused key
+  if (focusedRowIndex === -1) {
+    return 0;
+  }
+
+  const focusedElement = document.activeElement;
+  if (!focusedElement) {
+    return 0;
+  }
+
+  const { x, width } = focusedElement.getBoundingClientRect();
+  const targetX = Math.floor(x + width / 2); // Midpoint of button
+
+  const adjacentRowIndex = getAdjacentRowIndex(
+    keyMap,
+    focusedRowIndex,
+    direction
+  );
+  const adjacentRow = rowRefs.current[adjacentRowIndex];
+  if (!adjacentRow) {
+    return 0;
+  }
+
+  const adjacentRowKeys = Array.from(adjacentRow.querySelectorAll('button'));
+  const closestEdges = adjacentRowKeys.map((button) => {
+    const { x: currentX, width: currentWidth } = button.getBoundingClientRect();
+    const leftEdge = currentX;
+    const rightEdge = currentX + currentWidth;
+    const leftEdgeDistance = Math.abs(leftEdge - targetX);
+    const rightEdgeDistance = Math.abs(rightEdge - targetX);
+    return Math.min(leftEdgeDistance, rightEdgeDistance);
+  });
+
+  return closestEdges.indexOf(Math.min(...closestEdges));
+}
+
+/**
+ * Returns the index of the Key to focus if the user navigates to the previous row.
+ * Index of the Key is relative to its parent row only (not to all keys on the keyboard).
+ */
+function getPrevRowTargetButtonIndex(
+  keyMap: KeyMap,
+  focusedRowIndex: number,
+  rowRefs: React.MutableRefObject<Array<HTMLDivElement | null>>
+) {
+  return getAdjacentRowTargetButtonIndex(keyMap, rowRefs, focusedRowIndex, -1);
+}
+
+/**
+ * Returns the index of the Key to focus if the user navigates to the next row.
+ * Index of the Key is relative to its parent row only (not to all keys on the keyboard).
+ */
+function getNextRowTargetButtonIndex(
+  keyMap: KeyMap,
+  focusedRowIndex: number,
+  rowRefs: React.MutableRefObject<Array<HTMLDivElement | null>>
+) {
+  return getAdjacentRowTargetButtonIndex(keyMap, rowRefs, focusedRowIndex, 1);
+}
+
 export function VirtualKeyboard({
   onBackspace,
   onKeyPress,
+  onCancel,
+  onAccept,
   keyDisabled,
   keyMap = US_ENGLISH_KEYMAP,
+  enableWriteInAtiControllerNavigation,
 }: VirtualKeyboardProps): JSX.Element {
+  const [focusedRowIndex, setFocusedRowIndex] = useState(-1);
+
+  const keyMapWithActions: KeyMap = useMemo(() => {
+    const actions = enableWriteInAtiControllerNavigation
+      ? [
+          [SPACE_BAR_KEY, DELETE_KEY],
+          [CANCEL_KEY, ACCEPT_KEY],
+        ]
+      : // Cancel and Accept keys are rendered outside this component when ATI Controller navigation is off
```

Comment:
Any reason not to just always render them in this component? Seems like it would remove some complexity

Reply:
Mostly wanted to avoid reimplementing

1. action keys in ScanPanelVirtualKeyboard (the sip & puff keyboard)
2. the landscape-orientation sidebar for action keys in VxMark v2